File: api/src/main/java/org/openmrs/BaseDosingInstructions.java
/**
 * This Source Code Form is subject to the terms of the Mozilla Public License,
 * v. 2.0. If a copy of the MPL was not distributed with this file, You can
 * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
 * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
 *
 * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
 * graphic logo is a trademark of OpenMRS Inc.
 */
package org.openmrs;

import java.util.Date;

import static org.apache.commons.lang3.time.DateUtils.addSeconds;

/**
 * @since 2.3.4
 */
public abstract class BaseDosingInstructions implements DosingInstructions {

	/**
	 * @see DosingInstructions#getAutoExpireDate(DrugOrder)
	 */
	@Override
	public Date getAutoExpireDate(DrugOrder drugOrder) {
		if (drugOrder.getDuration() == null || drugOrder.getDurationUnits() == null) {
			return null;
		}
		String durationCode = Duration.getCode(drugOrder.getDurationUnits());
		if (durationCode == null) {
			return null;
		}
		Duration duration = new Duration(drugOrder.getDuration(), durationCode);
		return aMomentBefore(duration.addToDate(drugOrder.getEffectiveStartDate(), drugOrder.getFrequency()));
	}

	private Date aMomentBefore(Date date) {
		return addSeconds(date, -1);
	}
}

File: api/src/main/kotlin/org/openmrs/BaseDosingInstructions.kt
/**
 * This Source Code Form is subject to the terms of the Mozilla Public License,
 * v. 2.0. If a copy of the MPL was not distributed with this file, You can
 * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
 * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
 *
 * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
 * graphic logo is a trademark of OpenMRS Inc.
 */
package org.openmrs

import org.apache.commons.lang3.time.DateUtils.addSeconds
import java.util.Date

/**
 * @since 2.3.4
 */
abstract class BaseDosingInstructions : DosingInstructions {

	/**
	 * @see DosingInstructions.getAutoExpireDate
	 */
	override fun getAutoExpireDate(drugOrder: DrugOrder): Date? {
		if (drugOrder.duration == null || drugOrder.durationUnits == null) {
			return null
		}
		val durationCode = Duration.getCode(drugOrder.durationUnits) ?: return null
		val duration = Duration(drugOrder.duration, durationCode)
		return aMomentBefore(duration.addToDate(drugOrder.effectiveStartDate, drugOrder.frequency))
	}

	private fun aMomentBefore(date: Date?): Date? {
		return date?.let { addSeconds(it, -1) }
	}
}
